src/entity/effect/EffectContainer.php

```diff
@@ -0,0 +1,239 @@
+	 */
+	protected \Closure $effectValidatorForBubbles;
+
+	public function __construct(){
+		$this->defaultBubbleColor = new Color(0, 0, 0, 0);
+		$this->effectAddHooks = new ObjectSet();
+		$this->effectRemoveHooks = new ObjectSet();
+
+		$this->setEffectValidatorForBubbles(function(EffectInstance $effect) : bool{
+			return $effect->isVisible() && $effect->getType()->hasBubbles();
+		});
+	}
+
+	/**
+	 * Returns an array of Effects currently active.
+	 * @return EffectInstance[]
+	 */
+	public function all() : array{
+		return $this->effects;
+	}
+
+	/**
+	 * Removes all effects.
+	 */
+	public function clear() : void{
+		foreach($this->effects as $effect){
+			$this->remove($effect->getType());
+		}
+	}
+
+	/**
+	 * Removes the effect with the specified ID.
+	 */
+	public function remove(Effect $effectType) : void{
+		$index = spl_object_id($effectType);
+		if(isset($this->effects[$index])){
+			$effect = $this->effects[$index];
+
+			unset($this->effects[$index]);
+			foreach($this->effectRemoveHooks as $hook){
+				$hook($effect);
+			}
+
+			$this->recalculateEffectColor();
+		}
+	}
+
+	/**
+	 * Returns the effect instance active with the specified ID, or null if does not have the
+	 * effect.
+	 */
+	public function get(Effect $effect) : ?EffectInstance{
+		return $this->effects[spl_object_id($effect)] ?? null;
+	}
+
+	/**
+	 * Returns whether the specified effect is active.
+	 */
+	public function has(Effect $effect) : bool{
+		return isset($this->effects[spl_object_id($effect)]);
+	}
+
+	/**
+	 * In the following cases it will return true:
+	 * If a weaker effect of the same type is already applied.
+	 * If a weaker or equal-strength effect is already applied but has a shorter duration.
+	 *
+	 * @return bool whether the effect can been applied.
+	 */
+	public function canAdd(EffectInstance $effect) : bool{
+		$index = spl_object_id($effect->getType());
+		if(isset($this->effects[$index])){
+			$oldEffect = $this->effects[$index];
+			if(
+				abs($effect->getAmplifier()) < $oldEffect->getAmplifier()
+				|| (abs($effect->getAmplifier()) === abs($oldEffect->getAmplifier()) && $effect->getDuration() < $oldEffect->getDuration())
+			){
+				return false;
+			}
+		}
+		return true;
+	}
+
+	/**
+	 * Adds an effect.
+	 * If {@link EffectContainer::canAdd()} conditions are met.
+	 *
+	 * @return bool whether the effect has been successfully applied.
+	 */
+	public function add(EffectInstance $effect) : bool{
+		if($this->canAdd($effect)){
+			$index = spl_object_id($effect->getType());
+			$replacesOldEffect = isset($this->effects[$index]);
+
+			$this->effects[$index] = $effect;
+			foreach($this->effectAddHooks as $hook){
+				$hook($effect, $replacesOldEffect);
+			}
+
+			$this->recalculateEffectColor();
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * @phpstan-param \Closure(EffectInstance) : bool $effectValidator
+	 */
+	public function setEffectValidatorForBubbles(\Closure $effectValidator) : void{
```

Comment:
I think this would be better named "effect filter" rather than "effect validator".